### PR TITLE
Recreate managed resource secrets that have deletion timestamp

### PR DESCRIPTION
**Pull request by dimityrmirchev**
**How to categorize this PR?**

/area quality
/kind cleanup

**What this PR does / why we need it**:
In https://github.com/gardener/gardener/pull/8398 a fix for unwanted garbage collection of MR secrets was provided, but in some systems there are still such secrets that have deletion timestamp. After https://github.com/gardener/gardener/pull/8745 the finalizer that protects them will be gone. This PR introduces a recreation step so gardenlet can recreate these secrets on startup.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
/invite @rfranzke 

Thanks for reporting!

**Release note**:

```other operator
NONE
```

### CODE REVIEW

File: cmd/gardenlet/app/app.go [hunk 1]
```diff
@@ -480,6 +489,142 @@ func (g *garden) cleanupOrphanedExtensionsServiceAccounts(ctx context.Context, g
 	return flow.Parallel(taskFns...)(ctx)
 }
 
+const (
+	grmFinalizer           = "resources.gardener.cloud/gardener-resource-manager"
+	tempSecretLabel        = "resources.gardener.cloud/temp-secret"
+	tempSecretOldNameLabel = "resources.gardener.cloud/temp-secret-old-name"
+)
+
+// TODO(dimityrmirchev): Remove this code after v1.84 has been released.
```
Comment (rfranzke) [inline cmd/gardenlet/app/app.go]: To be on the safe side, let's keep it for a little longer (usually we do 3 releases)

```suggestion
// TODO(dimityrmirchev): Remove this code after v1.87 has been released.
```

File: cmd/gardenlet/app/app.go [hunk 2]
```diff
@@ -480,6 +489,142 @@ func (g *garden) cleanupOrphanedExtensionsServiceAccounts(ctx context.Context, g
...
+			return err
+		}
+	}
+	return nil
+}
+
+// TODO(dimityrmirchev): Remove this code after v1.84 has been released.
+func removeFinalizersAndWait(ctx context.Context, c client.Client, secret *corev1.Secret) error {
+	patch := client.StrategicMergeFrom(secret.DeepCopy())
+	secret.Finalizers = []string{}
```
Comment (rfranzke) [inline cmd/gardenlet/app/app.go:621]:
```suggestion
	secret.Finalizers = nil
```

File: cmd/gardenlet/app/app.go [hunk 3]
```diff
@@ -480,6 +489,142 @@ func (g *garden) cleanupOrphanedExtensionsServiceAccounts(ctx context.Context, g
...
+	if err := c.Patch(ctx, secret, patch); err != nil {
+		return fmt.Errorf("failed to patch the original secret %w", err)
+	}
+
+	goneSecret := &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: secret.Name, Namespace: secret.Namespace}}
```
Comment (rfranzke) [inline cmd/gardenlet/app/app.go]: Why is this needed? You could just continue using `secret` in the `WaitUntilResourceDeleted` call?

File: cmd/gardenlet/app/app.go [hunk 4]
```diff
@@ -480,6 +489,142 @@ func (g *garden) cleanupOrphanedExtensionsServiceAccounts(ctx context.Context, g
...
+				delete(original.Labels, tempSecretOldNameLabel)
+				original.ResourceVersion = ""
+				original.Name = originalName
+
+				if err := c.Create(ctx, original); err != nil {
+					return fmt.Errorf("failed to recreate the original secret %w", err)
+				}
+				if err := c.Delete(ctx, &temp); client.IgnoreNotFound(err) != nil {
+					return err
+				}
```
Comment (rfranzke) [inline cmd/gardenlet/app/app.go]: Put this logic into a function that is added to a list of `[]flow.Task`

File: cmd/gardenlet/app/app.go [hunk 5]
```diff
@@ -480,6 +489,142 @@ func (g *garden) cleanupOrphanedExtensionsServiceAccounts(ctx context.Context, g
...
+			if err := c.Create(ctx, original); err != nil {
+				return fmt.Errorf("failed to recreate the original secret %w", err)
+			}
+		}
+
+		// secret was already recreated. just delete the temporary one
+		if err := c.Delete(ctx, &temp); client.IgnoreNotFound(err) != nil {
+			return err
+		}
+	}
```
Comment (rfranzke) [inline cmd/gardenlet/app/app.go]: Put this logic into a function that is added to a list of `[]flow.Task`

File: cmd/gardenlet/app/app.go [hunk 6]
```diff
@@ -480,6 +489,142 @@ func (g *garden) cleanupOrphanedExtensionsServiceAccounts(ctx context.Context, g
...
+		original.ResourceVersion = ""
+		original.Finalizers = []string{}
+
+		// recreate the original and delete the temporary one
+		if err := c.Create(ctx, &original); err != nil {
+			return fmt.Errorf("failed to recreate the original secret %w", err)
+		}
+		if err := c.Delete(ctx, tempSecret); client.IgnoreNotFound(err) != nil {
+			return err
+		}
```
Comment (rfranzke) [inline cmd/gardenlet/app/app.go]: Put this logic into a function that is added to a list of `[]flow.Task`

File: cmd/gardenlet/app/app.go [hunk 7]
```diff
@@ -480,6 +489,142 @@ func (g *garden) cleanupOrphanedExtensionsServiceAccounts(ctx context.Context, g
...
+	}
+	return nil
```
Comment (rfranzke) [inline cmd/gardenlet/app/app.go]: Execute all collected `[]flow.Task`s here in parallel

File: cmd/gardenlet/app/app.go [hunk 8]
```diff
@@ -480,6 +489,142 @@ func (g *garden) cleanupOrphanedExtensionsServiceAccounts(ctx context.Context, g
...
+}
+
+// TODO(dimityrmirchev): Remove this code after v1.84 has been released.
```
Comment (rfranzke) [inline cmd/gardenlet/app/app.go]: 87

File: cmd/gardenlet/app/app.go [hunk 9]
```diff
@@ -480,6 +489,142 @@ func (g *garden) cleanupOrphanedExtensionsServiceAccounts(ctx context.Context, g
...
+		return fmt.Errorf("failed to patch the original secret %w", err)
+	}
+
+	goneSecret := &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: secret.Name, Namespace: secret.Namespace}}
+	cancelCtx, cancel := context.WithTimeout(ctx, time.Second*10)
+	defer cancel()
+	return kubernetesutils.WaitUntilResourceDeleted(cancelCtx, c, goneSecret, 1*time.Second)
+}
+
+// TODO(dimityrmirchev): Remove this code after v1.84 has been released.
```
Comment (rfranzke) [inline cmd/gardenlet/app/app.go]: 87

File: cmd/gardenlet/app/app.go [hunk 10]
```diff
@@ -480,6 +489,142 @@ func (g *garden) cleanupOrphanedExtensionsServiceAccounts(ctx context.Context, g
...
+func getSecretsToRecreate(ctx context.Context, c client.Client) ([]corev1.Secret, error) {
+	selector := labels.NewSelector()
+	isGC, err := labels.NewRequirement(references.LabelKeyGarbageCollectable, selection.Equals, []string{"true"})
+	if err != nil {
+		return nil, err
+	}
+	notTemp, err := labels.NewRequirement(tempSecretLabel, selection.DoesNotExist, []string{})
```
Comment (rfranzke) [inline cmd/gardenlet/app/app.go]:
```suggestion
	notTemp, err := labels.NewRequirement(tempSecretLabel, selection.DoesNotExist, nil)
```

File: cmd/gardenlet/app/app.go [hunk 11]
```diff
@@ -480,6 +489,142 @@ func (g *garden) cleanupOrphanedExtensionsServiceAccounts(ctx context.Context, g
...
+	if err != nil {
+		return nil, err
+	}
+	selector.Add(*isGC, *notTemp)
+	secretList := &corev1.SecretList{}
+	if err := c.List(ctx, secretList, client.MatchingLabelsSelector{Selector: selector}); err != nil {
+		return nil, err
+	}
+	secretsToRecreate := slices.DeleteFunc(secretList.Items, func(s corev1.Secret) bool {
+		return !slices.Contains(s.Finalizers, grmFinalizer)
```
Comment (rfranzke) [inline cmd/gardenlet/app/app.go]: What about the deletion timestamp? Aren't you now also considering secrets which DON'T have a deletion timestamp but just the finalizer (which is perfectly fine)?

File: cmd/gardenlet/app/app.go [hunk 12]
```diff
@@ -480,6 +489,142 @@ func (g *garden) cleanupOrphanedExtensionsServiceAccounts(ctx context.Context, g
...
+	}
+
+	secretsToRecreate, err := getSecretsToRecreate(ctx, c)
+	if err != nil {
+		return fmt.Errorf("failed listing secrets for recreation %w", err)
+	}
+
+	for _, original := range secretsToRecreate {
+		tempSecret := original.DeepCopy()
+		tempSecret.Name = original.Name + "-tmp"
```
Comment (rfranzke) [inline cmd/gardenlet/app/app.go]: I would rather prefix it with `tmp-` for easier grouping

File: cmd/gardenlet/app/app.go [hunk 13]
```diff
@@ -480,6 +489,142 @@ func (g *garden) cleanupOrphanedExtensionsServiceAccounts(ctx context.Context, g
...
+		metav1.SetMetaDataLabel(&tempSecret.ObjectMeta, tempSecretLabel, "true")
+		metav1.SetMetaDataLabel(&tempSecret.ObjectMeta, tempSecretOldNameLabel, original.Name)
+		tempSecret.DeletionTimestamp = nil
+		tempSecret.ResourceVersion = ""
+		tempSecret.Finalizers = []string{}
```
Comment (rfranzke) [inline cmd/gardenlet/app/app.go]:
```suggestion
		tempSecret.Finalizers = nil
```

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: cmd/gardenlet/app/app.go (hunk 3 above)
Comment (dimityrmirchev) [reply cmd/gardenlet/app/app.go]: `WaitUntilResourceDeleted` uses the passed object and continuously modifies it via `Get` calls. The initial thought was to not modify the passed secret to this function. Either way it is called with argument `original.DeepCopy` so I guess we can apply your suggestion.

File: cmd/gardenlet/app/app.go (hunk 11 above)
Comment (dimityrmirchev) [reply cmd/gardenlet/app/app.go]: Yes, you are right. Should be fixed with https://github.com/gardener/gardener/pull/8812/commits/0d25953b938b04a0c37e401f6e286efb5fd16a67

**Comment by dimityrmirchev** [comment]
@rfranzke thanks for the review. I implemented most of the suggestions. My only concern with the `flow.Parallel()` is that these can be many (possibly in the hundreds?) secrets that need to be recreated. I see that the current implementation does not provide a setting to set the degree of parallelism. Wouldn't hundreds of goroutines making calls to the API server cause client or server side throttling in which case we might observe failures?

**Comment by rfranzke** [comment]
> @rfranzke thanks for the review. I implemented most of the suggestions. My only concern with the `flow.Parallel()` is that these can be many (possibly in the hundreds?) secrets that need to be recreated. I see that the current implementation does not provide a setting to set the degree of parallelism. Wouldn't hundreds of goroutines making calls to the API server cause client or server side throttling in which case we might observe failures?

Well, I assume that the API server can properly handle a few hundred create/update requests to secrets. This is a one time action anyways and only affects the seed API server and nothing else. I wouldn't be so afraid. These operations are fast and simple (no continuous listing required, etc.). If you still want to limit it, you can follow an easy approach like in https://github.com/gardener/gardener/blob/master/pkg/utils/gardener/secretsrotation/etcd.go#L105-L136 with the `rate.NewLimiter` and then `limiter.Wait` in the flow task function, but I think it's overkill. For sure, you should introduce the parallelism, please.

**Comment by dimityrmirchev** [comment]
@rfranzke Thanks for the `rate.Limiter` hint. I think that the concerns should be addressed now. PTAL

**Review by rfranzke — COMMENTED** [review]
/lgtm
/approve
